### DISCUSSION TITLE
How to install a specified version of MOOSE?

### DISCUSSION THREAD

**Discussion by LuMeHa12138** [idaholab/moose#24895] [Q&A Getting Started] [answered]
## Reason
Due to certain special reasons, I would like to install a specific version of MOOSE, especially the old version (with modified InputParameter template).

I searched some issues and found the Issue #21548.
The questioner attempted to use the git command to return the version but encountered some problems. I sent a private message to him, and he said that he still hasn't solved the problem.
And I also found there are several versions in the Release Page, but I can not find some install tutorials about them.
So I still have not a good idea about it, and I hope the official can provide relevant tutorials for the users, please!
If this type of tutorial needs to be tailored to different versions, please prioritize providing some guidance for that version snapshot-20-10-27.

Thank you!!!

**Comment by GiudGiud** [comment]
hello

A discussions post is more appropriate here, I converted it.

You can do this easily by checking out an earlier commit with `git checkout <commit_hash>`
Do you have a link to this branch? snapshot-20-10-27
I can advise you on the commit from there. 

Note that we do not recommend doing this at all. You will miss out on years of bug fixes.

Guillaume

**Reply by LuMeHa12138** [reply]
Thank you for your answer. I took your advice. So I took these steps as follow:

1. (moose_old) lmh@ubuntu:~/projects$ git clone https://github.com/idaholab/moose.git
2. cd moose && git checkout b57161c
3. git clean -nxfd && git clean -xfd && scripts/update_and_rebuild_libmesh.sh
It prompts that  “If you did not run the update_and_rebuild_petsc.sh script yet, please run it before building libMesh”
So I took the next.
4. scripts/update_and_rebuild_petsc.sh
It prompts that "python3 is missing distutils.sysconfig, Skipping...
Error! Could not locate Python (with distutils.sysconfig)! Please install and have python3, python, or python2 (with distutils.sysconfig) in PATH."
So I took the next
5. sudo apt install python3-pip && sudo pip install disutils
6. scripts/update_and_rebuild_petsc.sh
In the end, it prompts a configure mistake about petsc. I put this log file here.
Could you please give me some guidance about it?

[configure.log](https://github.com/idaholab/moose/files/11947705/configure.log)

**Reply by GiudGiud** [reply]
```
'InShell' object has no attribute 'isAlive'  File "/home/lmh/projects/moose/petsc/config/configure.py", line 451, in petsc_configure
```

See this post
https://github.com/jupyter-vim/jupyter-vim/issues/51

you ll need to go rename this call from isAlive to is_alive.
And this is the first of many issues you'll hit by using an old version

what's the issue with using the current MOOSE?

**Reply by milljm** [reply, marked as answer]
It's entirely possible GCC 11.3.0 (released Apr 2022) is too new a compiler to build this hash of MOOSE. I know this, as we have ran into GCC 11.x issues in the past (a year or so ago).

Ubuntu 22.04 LTS comes with 11.2 (I am not willing to install a non-LTS version, so I will have to stick with 22.04). I am assuming you are using Ubuntu 22.10 or 23.04 (non-LTS versions) (reference: https://wiki.ubuntu.com/Releases)? I am only curious if we have to delve into this even more.

I am making points about the version of GCC you are using due to the non-presence of `mpicc` as described below. I am not sure if you are wanting to use Conda, or your system compilers?


While setting up my environment, I realized the following occurred in your log:

```pre
❯ cat configure.log | grep mpicc
Checking for program /home/lmh/projects/moose/python/peacock/mpicc...not found
Checking for program /home/lmh/miniconda3/envs/moose_old/bin/mpicc...not found
Checking for program /home/lmh/miniconda3/condabin/mpicc...not found
Checking for program /usr/local/sbin/mpicc...not found
Checking for program /usr/local/bin/mpicc...not found
Checking for program /usr/sbin/mpicc...not found
Checking for program /usr/bin/mpicc...not found
Checking for program /sbin/mpicc...not found
Checking for program /bin/mpicc...not found
Checking for program /usr/games/mpicc...not found
Checking for program /usr/local/games/mpicc...not found
Checking for program /snap/bin/mpicc...not found
Checking for program /snap/bin/mpicc...not found
Checking for program /home/lmh/projects/moose/petsc/lib/petsc/bin/win32fe/mpicc...not found
  Unable to find programs ['mpicc'] providing listing of each search directory to help debug
```

My concerns:

```pre
Checking for program /home/lmh/miniconda3/envs/moose_old/bin/mpicc...not found
```

A `moose_old` Conda environment still lingers among your PATH. Which in and of itself is odd, that 1, it exists when it possibly should not? And 2, should exist and does not contain an `mpicc` binary. Maybe this directory has been deleted?

I think the best path forward, is to install the oldest possible `moose-mpich moose-tools` package we have, and go from there.

```bash
conda config --add channels https://conda.software.inl.gov/archive             # An older channel we do not advertise
mamba create -n moose_old moose-tools=2021.07.14 moose-mpich python=3.7 cmake  # overwrite `moose_old`
mamba activate old_mpich
export METHODS=opt METHOD=opt MOOSE_JOBS=6                                     #  to speed up the build process
```

This above will install an environment which should work with that older version of PETSc you are trying to build...

But as @GiudGiud asked, the absolute best option is to know why you cannot use the latest version of MOOSE?

**Reply by LuMeHa12138** [reply]
Thank you for your reply.
I have tried to complete the installation of this version of Moose in the past few days.
I would like to share the specific process.

1. I installed Python 3.7.6 in the Conda environment(moose_ old ), supplemented with gfortran, bison, and flex.
2. I ran the script to rebuild petsc using the following command:
./scripts/update_and_rebuild_petsc.sh -- download-mpich=1-- download-cmake=1-- with-fc-dir=/usr/bin/gfortran
3. I ran the script to rebuild libmesh:
./scripts/update_and_rebuild_ libmesh.sh

Everything runs well, although there are too many moose tests failed to trigger a warning. I still successfully compiled my previous MooseApp in this version of Moose environment as expected and ran through the example to obtain similar results.

The following are my specific moose environment version for reference.
Framework Information:
MOOSE Version:    git commit b57161c312 on 2020-10-24
LibMesh Version:   f33331f6ebb0f31dff7526106268ca0c4300049e
PETSc Version:      3.13.3
SLEPc Version:      3.13.3

Next, I will answer your question (why not use the latest moose). I have been developing my own MooseApp in the old version of Moose environment before. Now I want to install the same version of the environment in a new location to run my MooseApp.

In the end, I am curious about some methods you mentioned for installing older versions. Can you provide some pages for placing relevant content on the Moose official website or open up those old installation tutorial pages for reference.

Thank you！

**Reply by milljm** [reply]
- ./scripts/update_and_rebuild_petsc.sh -- download-mpich=1

works for PETSc, but this will not get MPICH for anything except PETSc. You need to provide a system-wide version of MPI instead, so MPI becomes available to libMesh, MOOSE, etc.

Everything will build fine, because libMesh does not require MPI in order to be built. While actually running MOOSE does. So, the errors you are getting when running tests (running moose), are expected.

Ironically, later versions of these scripts your are running have sanity checks in them to alert you, in the event you try to build libMesh without MPI.


Creating instructional pages on how to use older versions of MOOSE is a good idea, but would be problematic. We can only go back so far, and unfortunately in your case, you're going back far enough where we were just beginning to use Conda. We made a lot of mistakes creating those packages back then. So we moved them out from the public channel, into a separate 'archive' channel. That is: they cannot coexist in the same public channel (https://conda.software.inl.gov/public). Doing so causes all sorts of conflicting headaches. Basically what I am trying to say is, even if we created instructions on how to use older versions of MOOSE, the version you want would be beyond the scope of what we'd support showing folks how to do.

For example, in your case, you need to install a version of Python that has reached end of life (https://endoflife.date/python 3.7 is now EOL as of the end of June). When this happens, some maintainers will go so far as to **_physically remove packages from Conda-Forge_** (something we cannot prevent).

Going through the instructions I wrote earlier, it would seem all the necessary Conda packages are still available, and everything builds fine for me:

```bash
(testing) [ubuntu-22][~/projects/3rd/moose/test]> ./moose_test-opt -h
SLEPc Release Version 3.13.3, unknown
       The SLEPc Team
    slepc-maint@upv.es
 https://slepc.upv.es
See docs/manual.html for help.
<trimmed>


(testing) [ubuntu-22][~/projects/3rd/moose/test]> git rev-parse HEAD
b57161c312e90a33cf73fcf5f8ce9c0191af6155


(testing) [ubuntu-22][~/projects/3rd/moose/test]> mamba list | grep moose
moose-mpich               3.3.2                   build_7    https://conda.software.inl.gov/archive
moose-tools               2021.07.14       py37hedbaeeb_0    https://conda.software.inl.gov/archive


(testing) [ubuntu-22][~/projects/3rd/moose/test]> ./run_tests -j 12
dgkernels/2d_diffusion_dg.no_additional_rms_distributed ........................ [MESH_MODE!=DISTRIBUTED] SKIP
dgkernels/2d_diffusion_dg.proper_ghosting_with_action_serial_distributed ....... [MESH_MODE!=DISTRIBUTED] SKIP
dgkernels/2d_diffusion_dg.proper_ghosting_with_action_parallel_distributed ..... [MESH_MODE!=DISTRIBUTED] SKIP

<trimmed>

materials/derivative_material_interface.ad_mutliblock ..................................................... OK
materials/derivative_material_interface.ad_warn ........................................................... OK
--------------------------------------------------------------------------------------------------------------
Ran 2395 tests in 184.5 seconds.
2395 passed, 58 skipped, 0 pending, 0 failed
(testing) [ubuntu-22][~/projects/3rd/moose/test]> 
```
